File: gffFormater.py
__author__ = 'johngiorgi'

import collections

############## Classes ##############
class Gene:

    def __init__(self, name, listOfExons, listOfCDS, derivedLine):
        self.name = name
        self.listOfExons = listOfExons
        self.listOfCDS = listOfCDS
        self.derivedLine = derivedLine


    def get_listOfExons(self):
        return self.listOfExons
    def get_listOfCDS(self):
        return self.listOfCDS
    def get_derivedLine(self):
        return self.derivedLine
############## Classes ##############


def formatGFF(input,output):
    '''
    This function will loop through every line of a GFF file, creating a new gene object in memory which will
    group all associated exons and CDS. This is determined by the exon and CDS objects "name" attribute. Finally,
    the function will deconstruct the information from the memory structure to create a new, properly formatted
    GFF file, where the hierarchy Gene > mRNA > exon > CDS is observed for every gene.

    Note: This function requires that the GFF follows the GFF3 sequence ontology formatting. What it will fix
    is issues with the hierarchical structure, i.e. not following a Gene > all of this genes mRNA > all of this
    genes Exons > all of this genes CDS layout. 

    :param input: directory path to the GFF file to be formatted
    :param output: directory path to the location to save a NEW formatted GFF
    '''

    # A ordered dictionary to store our genes
    geneDictionary = collections.OrderedDict()
    # A list of names
    geneNames = []


    # Store every line of the input GFF in an array, data
    with open(input, "r") as inputFile:
        data = inputFile.readlines()

    # Our loop condition. We want all lines in data that do not have ##gff-version in them
    loopCondition = (line for line in data if "##gff-version" not in line)

    '''(1) MEMORY: this loop will create our gene objects in memory, creating a hierarchical pattern (gene>exon>CDS)'''

    for line in loopCondition:

        # Get the name of the gene
        geneName = (((line.split("\t"))[8]).split(";"))[1]

        # If we have not already created a gene object of name "geneName" then
        # append a new Gene object to listOfGenes with name geneName.
        if geneName not in geneNames:

            # Create our "tempGene" to store the gene we are working with,
            # and append it to our gene dictionary
            tempGene = Gene(geneName, [], [], line)
            geneDictionary[geneName] = tempGene
            # Store the name of this gene
            geneNames.append(geneName)

        # Otherwise, if we have seen this name before, make our tempGene
        # equal to our already created gene object by accessing it from the
        # dictionary
        else:
            tempGene = geneDictionary[geneName]

        # Determine if the line represents an exon or CDS
        exonOrCDS = (line.split("\t"))[2]

        if exonOrCDS == "exon":
            # Append the line to the listOfExons in the gene object
            (tempGene.get_listOfExons()).append(line)
        elif exonOrCDS == "CDS":
            # Append the line to the listOfCDS in the gene object
            (tempGene.get_listOfCDS()).append(line)

    '''(2) FILE: this loop will deconstruct our gene objects in memory, producing a properly formatted GFF file'''

    with open(output, "w") as outputFile:
        for name in geneNames:

            # Our current gene will be accessed from the ordered dictionary by its name
            currentGene = geneDictionary[name]

            # Then we will get the line the gene was derived from, and change
            # it's "type" to gene
            currentGenesLine = currentGene.get_derivedLine()
            currentGenesLine = currentGenesLine.split("\t")
            currentGenesLine[2] = "gene"

            # Determine the start and stop positions of the gene based on the first exon
            # This is used as a starting point
            start = (((currentGene.get_listOfExons())[0]).split('\t'))[3]
            stop = (((currentGene.get_listOfExons())[0]).split('\t'))[4]

            # The loop will move through all exons of our gene, finding the smallest "start" and largest "stop"
            # thus finding the start and stop positions at the level of the gene.
            for item in currentGene.get_listOfExons():
                tempStart = (item.split("\t"))[3]
                tempStop = (item.split("\t"))[4]
                if tempStart < start:
                    start = tempStart
                if tempStop > stop:
                    stop = tempStop

            # Assign the gene its start and stop positions.
            currentGenesLine[3] = start
            currentGenesLine[4] = stop

            # Join our array, this will represent the "gene" line to write to the GFF3 file.
            currentGenesLine = "\t".join(currentGenesLine)

            # Finally, write out gene line to the GFF.
            outputFile.write(currentGenesLine)


            # These two foor loops will write all exons, and then all CDS of our gene to the GFF3 file.
            for item in currentGene.get_listOfExons():
                outputFile.write(item)
            for item in currentGene.get_listOfCDS():
                outputFile.write(item)



formatGFF("inccorectlyFormated_FORMATFIXED.gff", "inccorectlyFormated_FORMATFIXED_HIERARCHYFIXED.gff")

File: gffFormater.go
/*
* Work on parallel processing
*
*
 */
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
)

func main() {

	// Printing to the command line
	// fmt.Println(buildCorrectedGFF("incorrectFormat.gff"))

	// Call buildCorrectedGFF
	buildCorrectedGFF("inccorectlyFormated.gff")
}

/*
* This function will attempt to correct a given GFF files format, according to the GFF3 specifications maintained by The Sequence Oncology Project
* @params: a string representing the path for the gff file to be formatted
*/
func buildCorrectedGFF(path string) {
	// Run function readlines to read in data
	data, err := readLines(path)

	// Print error if it arises
	if err != nil {
		fmt.Println(err)
	}

	// A slice to hold our corrected data
	var correctedLines []string
	// Counter to give lines unique ID's if neccecary
	IDCounter := 0

	// A for each loop to move through each line of the GFF
	for _, lines := range data {

		// Split each line at its tab
		splitAtTab := strings.Split(lines, "\t")

		// Stores the element of each line
		element := splitAtTab[2]

		// Iterate through ever line. Save each one in memory that is not a "start_codon" or "stop_codon" item.
		if element == "gene" || element == "CDS" || element == "exon" || element == "mRNA"|| element == "UTR" || element == "rRNA" {

			// Split up the attributes at each semicolon
			attributes := strings.Split(splitAtTab[8], "; ")

			// Add back in the semicolon for each attribute
			for index, attribute := range attributes {
				attributes[index] = attribute + ";"
			}

			
			/*THIS WAS WRITTEN LIKE THIS TO SAVE TIME. ASSUMES THAT THE GFF HAS AN INCORRECTLY FORMATED NAME ATTRIBUTE*/
			/*TO HANDLE MORE CASES, IF STATEMENTS LIKE THE ONE BELOW THAT CHECKS FOR AN ID SHOULD BE WRITTEN*/

			// Fixes the name attribute
			name_attribute := strings.Split(attributes[0], " ")
			attributes[0] = "Name=" + name_attribute[1]


			// Checks if attributes have an ID. If not prepends it to attributes
			if !(stringInSlice("ID", attributes)) {
				// Need a counter for each of the element types
				IDCounter += 1
				// Prepend unique ID to attributes
				attributes = append([]string{"ID=" + element + strconv.Itoa(IDCounter) +";"}, attributes...)
			}


			// Loops through attributes and replaces all spaces with = signs
			for index, _ := range attributes {
				attributes[index] = strings.Replace(attributes[index], " ", "=", 1)
			}
			
			/*THIS IS WHERE THE CORRECTED LINES ARE APPENED TO A SLICE*/

			// This will append each line of data, from coloums 0 to 7, and then the fixed attributes, followed by a newline, to corrected lines
			correctedLines = append(correctedLines, strings.Join(append(splitAtTab[:8]), "\t"))
			correctedLines = append(correctedLines,"\t")
			correctedLines = append(correctedLines, attributes...)
			correctedLines = append(correctedLines,"\n")


		} else {fmt.Println("Line element does not follow GFF3 convention: \n" + lines)}
		
	}

	// If the GFF file has no GFF version, this will append a directive to the beginning of the file.
	if !(strings.Contains(correctedLines[0], "##gff-version")) {
	correctedLines = append([]string{"##gff-version\t3\n"}, correctedLines...)
	}

	// Write the corrected lines to a file
	writeFile(path, correctedLines)
}


/*
* This function will return true if any elements in a input slice of strings, list, contains a substring, a, and false otherwise
* @params: a: substring to search for in the slice of strings
* @params list: a slice of strings to query
* @return true if any element of list[] contains string a.
*/

func stringInSlice(a string, list []string) bool {
    for _, b := range list {
        if strings.Contains(b,a) {
            return true
        }
    }
    return false
}

/*
 * This function will write the corrected GFF file, as returned by buildCorrectedGFF as a string splice, to a created file named 
 * <name of input file>_CORRECTED.gff" in the same directory the program was ran
 * @params: a string representing the path for the GFF file to be formatted
 * @return any error returned by the WriteString function.
 */
func writeFile(path string, sliceToWrite []string) (int, error) {
	
	// From the file path, produces a string with its name.gff changed to name_CORRECTED.gff
	// Savs this as newFilenameString
	newFilenameSlice := strings.Split(path, "/")
	newFilenameSlice = strings.Split(newFilenameSlice[len(newFilenameSlice)-1], ".")
	newFilenameSlice[0] = newFilenameSlice[0] + "_FORMATFIXED"
	newFilenameString := strings.Join(append(newFilenameSlice), ".")

	f, err := os.Create(newFilenameString)
	// Defer closing of the file
	defer f.Close()
	// Create new Writer object
	writer := bufio.NewWriter(f)
	// Write each line of the returned []string array containing the corrected lines of the GFF file at path
	n4, err := writer.WriteString(strings.Join(append(sliceToWrite), ""))
	return n4, err
}

/*
* Function will open GFF file and store it in memory
* @params: a string representing the path for the GFF file to be formatted
* @return: a string splice containing each line of the GFF file, and any errors reported by the compiler
 */
func readLines(path string) ([]string, error) {

	// Open the file
	file, err := os.Open(path)

	// Catch any erros
	if err != nil {
		return nil, err
	}

	// Close the file only if it was opened successfully and there are no errors
	defer file.Close()

	// Declare a splice to store each line of the GFF file
	var lines []string

	// Create new scanner object for the file
	scanner := bufio.NewScanner(file)

	// Scan the file using our scanner object. Append each line to splice lines
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}
